execution_engine/src/core/engine_state/upgrade.rs

```diff
@@ -184,37 +185,49 @@ where
     /// Creates new system upgrader instance.
     pub(crate) fn new(
         new_protocol_version: ProtocolVersion,
+        old_protocol_version: ProtocolVersion,
         tracking_copy: Rc<RefCell<TrackingCopy<<S as StateProvider>::Reader>>>,
     ) -> Self {
         SystemUpgrader {
             new_protocol_version,
+            old_protocol_version,
             tracking_copy,
         }
     }
 
-    /// Bump major version for system contracts.
-    pub(crate) fn upgrade_system_contracts_major_version(
+    /// Check if the next version is a major version bump.
+    pub(crate) fn is_major_bump(&self) -> bool {
+        self.old_protocol_version.value().major + 1 == self.new_protocol_version.value().major
+    }
+
+    /// Bump major version and/or update the entry points for system contracts.
+    pub(crate) fn refresh_system_contracts(
         &self,
         correlation_id: CorrelationId,
         mint_hash: &ContractHash,
         auction_hash: &ContractHash,
         handle_payment_hash: &ContractHash,
         standard_payment_hash: &ContractHash,
     ) -> Result<(), ProtocolUpgradeError> {
-        self.store_contract(correlation_id, *mint_hash, MINT, mint::mint_entry_points())?;
-        self.store_contract(
+        self.refresh_system_contract_entry_points(
```

Comment:
Let's pattern match on `SystemContractType` and handle all of its variants so that we have additional check that makes sure we handle all the cases.

Reply:
Using `SystemContractType`